// @dopex-io/sdk/src/types/typechain/factories/CurveGaugesOracle__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  CurveGaugesOracle,
  CurveGaugesOracleInterface,
} from "../CurveGaugesOracle";

const _abi = [
  {
    inputs: [
      {
        internalType: "address",
        name: "_gaugeSnapshotReceiverAddress",
        type: "address",
      },
      {
        internalType: "address",
        name: "_chainlinkCRVAddress",
        type: "address",
      },
    ],
    stateMutability: "nonpayable",
    type: "constructor",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "epochStart",
        type: "uint256",
      },
      {
        internalType: "uint256",
        name: "epochEnd",
        type: "uint256",
      },
      {
        internalType: "address",
        name: "gauge",
        type: "address",
      },
    ],
    name: "getRate",
    outputs: [
      {
        internalType: "uint256",
        name: "rate",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
];

export class CurveGaugesOracle__factory {
  static readonly abi = _abi;
  static createInterface(): CurveGaugesOracleInterface {
    return new utils.Interface(_abi) as CurveGaugesOracleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): CurveGaugesOracle {
    return new Contract(address, _abi, signerOrProvider) as CurveGaugesOracle;
  }
}
