// @dopex-io/sdk/src/types/typechain/factories/ETHVolatilityOracle__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Signer, utils } from "ethers";
import type { Provider } from "@ethersproject/providers";
import type {
  ETHVolatilityOracle,
  ETHVolatilityOracleInterface,
} from "../ETHVolatilityOracle";

const _abi = [
  {
    anonymous: false,
    inputs: [
      {
        indexed: true,
        internalType: "address",
        name: "previousOwner",
        type: "address",
      },
      {
        indexed: true,
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "OwnershipTransferred",
    type: "event",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "strike",
        type: "uint256",
      },
    ],
    name: "getVolatility",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "owner",
    outputs: [
      {
        internalType: "address",
        name: "",
        type: "address",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [],
    name: "renounceOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    name: "strikeToVols",
    outputs: [
      {
        internalType: "uint256",
        name: "",
        type: "uint256",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "address",
        name: "newOwner",
        type: "address",
      },
    ],
    name: "transferOwnership",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
  {
    inputs: [
      {
        internalType: "uint256[]",
        name: "strikes",
        type: "uint256[]",
      },
      {
        internalType: "uint256[]",
        name: "vols",
        type: "uint256[]",
      },
    ],
    name: "updateVolatility",
    outputs: [],
    stateMutability: "nonpayable",
    type: "function",
  },
];

export class ETHVolatilityOracle__factory {
  static readonly abi = _abi;
  static createInterface(): ETHVolatilityOracleInterface {
    return new utils.Interface(_abi) as ETHVolatilityOracleInterface;
  }
  static connect(
    address: string,
    signerOrProvider: Signer | Provider
  ): ETHVolatilityOracle {
    return new Contract(address, _abi, signerOrProvider) as ETHVolatilityOracle;
  }
}
